### ISSUE TITLE
SSL-SNI requirement causes failed voice websocket connects in Brazil region

### ISSUE THREAD

**Issue by ComputerMaster1st** [RogueException/Discord.Net#1064]
Hi There,

There appears to be connection difficulties with connecting to Brazilian Voice servers. I'm uncertain if this affects all D.Net powered bots at this time.

Here is the exception...

```
[ 23:38:12 ] <=> 23:38:12 Audio #1    Connecting
[ 23:38:13 ] <=> 23:38:13 Audio #1    System.Net.WebSockets.WebSocketException (0x80004005): Unable to connect to the remote server ---> System.Security.Authentication.AuthenticationException: The remote certificate is invalid according to the validation procedure.
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Net.Security.SslState.StartSendAuthResetSignal(ProtocolToken message, AsyncProtocolRequest asyncRequest, ExceptionDispatchInfo exception)
   at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.PartialFrameCallback(AsyncProtocolRequest asyncRequest)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Net.Security.SslState.InternalEndProcessAuthentication(LazyAsyncResult lazyResult)
   at System.Net.Security.SslState.EndProcessAuthentication(IAsyncResult result)
   at System.Net.Security.SslStream.EndAuthenticateAsClient(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.WebSockets.WebSocketHandle.<ConnectAsyncCore>d__24.MoveNext()
   at System.Net.WebSockets.WebSocketHandle.<ConnectAsyncCore>d__24.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.WebSockets.ClientWebSocket.<ConnectAsyncCore>d__16.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Discord.Net.WebSockets.DefaultWebSocketClient.<ConnectInternalAsync>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Discord.Net.WebSockets.DefaultWebSocketClient.<ConnectAsync>d__25.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Discord.Audio.DiscordVoiceAPIClient.<ConnectInternalAsync>d__58.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Discord.Audio.DiscordVoiceAPIClient.<ConnectInternalAsync>d__58.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Discord.Audio.DiscordVoiceAPIClient.<ConnectAsync>d__57.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Discord.Audio.AudioClient.<OnConnectingAsync>d__48.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Discord.ConnectionManager.<ConnectAsync>d__30.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Discord.ConnectionManager.<>c__DisplayClass28_0.<<StartAsync>b__0>d.MoveNext()
[ 23:38:13 ] <=> 23:38:13 Audio #1    Disconnecting
[ 23:38:13 ] <=> 23:38:13 Audio #1    Disconnected
```

Regards,

CM1

**Comment by FiniteReality** [comment]
This is likely due to SSL-SNI not working since Discord are rolling this out now. As per Discord API:

```
Voice Websockets SSL / SNI

We're slowly rolling out a planned upgrade to our voice infrastructure, and as a general FYI we'd like folks to be aware that the voice server websocket now requires a websocket client with a proper SSL-SNI implementation. So far I've only seen some issues crop up on a few Java libraries, but please check and confirm that your library implementations will work properly with this change. If you'd like to test, our Brazil voice region has already been upgraded with this change. Thanks :ablobsalute:
```

As simple debugging steps, have you tried using the latest System.Net.WebSockets? We might have to open an issue upstream if it still reproduces then.

**Comment by ComputerMaster1st** [comment]
Hi There,

Already running NETCore 2.0. I guess an upgrade is in order?

**Comment by FiniteReality** [comment]
.NET Core 2.0 doesn't include the latest System.Net.WebSockets - it's just a runtime. You can manually specify a System.Net.WebSockets dependency in your project, and if it's supported, I'm fairly sure Discord.Net will use it instead of the built-in one. If not, you might have to build Discord.Net yourself with the package version bumped.

**Comment by ComputerMaster1st** [comment]
Added System.Net.WebSockets 4.3.0 with no luck. Still kicking out security exception.

**Comment by FiniteReality** [comment]
I did some research and it looks like full SNI support is blocked behind .NET Core 2.1: https://github.com/dotnet/corefx/pull/28278

On Windows, it appears that SNI information is sent correctly however.

**Comment by ComputerMaster1st** [comment]
Guess I'll wait until something get's cooked up.

EDIT: Should I try compiling the full build of d.net with nugets updated on my end?

**Comment by FiniteReality** [comment]
You can definitely try it, but I doubt anything will work unless you use prerelease packages for .NET Core 2.1, and even then I don't guarantee that it'll work.

**Comment by ComputerMaster1st** [comment]
I'll try updating netcore first since I'm not running 2.1 yet. If not, then I'll try forking/cloning and updating nugets.

**Comment by ComputerMaster1st** [comment]
Issue solved! Turns out that NETCore 2.1 is now required. Took 3 hours to find & get the right version which allows targeting of net core 2.1. The way MS have done their version numbering is god awful.

Anyways, closing issue due to being resolved. :)

PS. dotnet-core-2.1.300-rc1 is the required install to allow netcore 2.1 targeting.

Thanks for letting me know Finite.
